src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -4,21 +4,105 @@
 using System.Collections.Concurrent;
 using System.Runtime.CompilerServices;
 using System.Threading.Channels;
+using Microsoft.Extensions.Hosting;
 using Microsoft.Extensions.Logging;
 
 namespace Aspire.Hosting.ApplicationModel;
 
 /// <summary>
 /// A service that allows publishing and subscribing to changes in the state of a resource.
 /// </summary>
-public class ResourceNotificationService(ILogger<ResourceNotificationService> logger)
+public class ResourceNotificationService
 {
     // Resource state is keyed by the resource and the unique name of the resource. This could be the name of the resource, or a replica ID.
     private readonly ConcurrentDictionary<(IResource, string), ResourceNotificationState> _resourceNotificationStates = new();
-    private readonly ILogger<ResourceNotificationService> _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+    private readonly ILogger<ResourceNotificationService> _logger;
+    private readonly CancellationToken _applicationStopping;
 
     private Action<ResourceEvent>? OnResourceUpdated { get; set; }
 
+    /// <summary>
+    /// Creates a new instance of <see cref="ResourceNotificationService"/>.
+    /// </summary>
+    /// <remarks>
+    /// Obsolete. Use the constructor that accepts an <see cref="ILogger{ResourceNotificationService}"/> and <see cref="IHostApplicationLifetime"/>.<br/>
+    /// This constructor will be removed in the next major version of Aspire.
+    /// </remarks>
+    /// <param name="logger">The logger.</param>
+    [Obsolete($"""
+        {nameof(ResourceNotificationService)} now requires an {nameof(IHostApplicationLifetime)}.
+        Use the constructor that accepts an {nameof(ILogger)}<{nameof(ResourceNotificationService)}> and {nameof(IHostApplicationLifetime)}.
+        This constructor will be removed in the next major version of Aspire.
+        """)]
+    public ResourceNotificationService(ILogger<ResourceNotificationService> logger)
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+    }
+
+    /// <summary>
+    /// Creates a new instance of <see cref="ResourceNotificationService"/>.
+    /// </summary>
+    /// <param name="logger">The logger.</param>
+    /// <param name="hostApplicationLifetime">The host application lifetime.</param>
+    public ResourceNotificationService(ILogger<ResourceNotificationService> logger, IHostApplicationLifetime hostApplicationLifetime)
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        _applicationStopping = hostApplicationLifetime?.ApplicationStopping ?? throw new ArgumentNullException(nameof(hostApplicationLifetime));
```

Comment:
When is this ever null?

Reply:
It was previously throwing the ANE for the logger parameter in the original constructor so I made it consistent when adding the new one. You want to remove bother ANEs?

Reply:
It's fine